### ISSUE TITLE
OTA flash failure when software UART receives data during flash

### ISSUE THREAD

**Issue by DAVe3283** [esphome/issues#2629]
### The problem

After updating from 2021.9.3 to 2021.10.1, the ESP8266 will reboot during an OTA flash if receiving data on a software UART.

* If I disconnect the software RX pin, the flash completes successfully.
* If I use the hardware UART, the flash completes successfully even when receiving UART data.
* Version 2021.9.3 does not have this problem. 2021.10.1 and 2021.10.2 do have the problem. 

If it matters, the data received is from a PMS7003 ([PMSX003 component](https://esphome.io/components/sensor/pmsx003.html)), but the reboot + OTA failure occurs with just a UART and no sensors defined. See YAML below-- that exact YAML will replicate the OTA crash.

In this application, we need 2 UARTs, so even if we swap the PMS7003 to the hardware UART, then the other component runs the risk of crashing an OTA if it sends an update.

### Which version of ESPHome has the issue?

2021.10.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2021.10.6

### What platform are you using?

ESP8266

### Board

D1 Mini

### Component causing the issue

UART

### Example YAML snippet

```yaml
esphome:
  name: haks-test
  platform: ESP8266
  board: d1_mini

logger:
  level: VERBOSE

api:
  password: "cd8c22eeaa2ae263b303aec49206103f"

ota:
  password: "cd8c22eeaa2ae263b303aec49206103f"

wifi:
  ssid: !secret iot_wifi_ssid
  password: !secret iot_wifi_pass
  domain: !secret iot_domain

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Haks-Test Fallback Hotspot"
    password: "cl766zRGasfA"

captive_portal:

uart:
  - id: uart_pms
    rx_pin: D5
    baud_rate: 9600
```

### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esphome/haks-test.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing haks-test (board: d1_mini; framework: arduino; platform: platformio/espressif8266 @ 2.6.2)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif8266 3.20704.0 (2.7.4) 
 - tool-esptool 1.413.0 (4.13) 
 - tool-esptoolpy 1.20800.0 (2.8.0) 
 - toolchain-xtensa 2.40802.200502 (4.8.2)
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.3
|   |-- <ESP8266WiFi> 1.0
|-- <ESPAsyncWebServer-esphome> 1.3.0
|   |-- <ESPAsyncTCP-esphome> 1.2.3
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <Hash> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS> 1.2
|   |-- <ESP8266WiFi> 1.0
|-- <noise-c> 0.1.3
|   |-- <libsodium> 1.10018.1
|-- <DNSServer> 1.1.1
|   |-- <ESP8266WiFi> 1.0
Compiling /data/haks-test/.pioenvs/haks-test/src/main.cpp.o
Linking /data/haks-test/.pioenvs/haks-test/firmware.elf
Retrieving maximum program size /data/haks-test/.pioenvs/haks-test/firmware.elf
Checking size /data/haks-test/.pioenvs/haks-test/firmware.elf
Building /data/haks-test/.pioenvs/haks-test/firmware.bin
RAM:   [====      ]  39.0% (used 31984 bytes from 81920 bytes)
Flash: [====      ]  39.1% (used 408184 bytes from 1044464 bytes)
Creating BIN file "/data/haks-test/.pioenvs/haks-test/firmware.bin" using "/data/cache/platformio/packages/framework-arduinoespressif8266/bootloaders/eboot/eboot.elf" and "/data/haks-test/.pioenvs/haks-test/firmware.elf"
========================= [SUCCESS] Took 4.24 seconds =========================
INFO Successfully compiled program.
INFO Resolving IP address of haks-test.****
INFO  -> 192.168.0.105
INFO Uploading /data/haks-test/.pioenvs/haks-test/firmware.bin (412336 bytes)
Uploading: [=========                                                   ] 15% 
ERROR Error sending data: [Errno 104] Connection reset by peer
```

**Comment by snh** [comment]
I too am experiencing this problem, in my case with a configuration similar to https://www.esphome-devices.com/devices/AirGradient-DIY.

**Comment by MallocArray** [comment]
Same here!  I'm doing the same project with AirGradient DIY with ESPHome and can flash it fine with USB, but if I try to update OTA, the device is rebooting mid flash.  I thought it may be due to the PMS5003 sending updates every second, so I tried to comment out the sensor, but I had left the UART section enabled.

Last night I downgraded esphome container to 1.18.00 and after flashing over USB I can now do OTA successfully. (Note: I did have to change my config file from .yml to .yaml for 1.18.0 to detect it.  

I thought it might be related to:
https://github.com/esphome/issues/issues/2309
as it see the same illegal instruction message, but I had tried disabling the captive portal and tweaking the output_power to 17.5db without any improvement on 2021.10.3

I have a very similar yaml file for the D1 Mini hardware setup mentioned by the OP

```yaml
esphome:
  name: airgradient-basement
  platform: ESP8266
  board: d1_mini

# Enable logging
logger:

# Enable Home Assistant API
api:
  password: ""

ota:
  password: ""

wifi:
  # https://esphome.io/components/wifi.html
  ssid: "MiWiFiFyi"
  password: "NachoPassword"
  output_power: 17.5db


  # Enable fallback hotspot (captive portal) in case wifi connection fails
#   ap:
#     ssid: "air-basement"
#     password: "asdf"
# captive_portal:

uart:
  # https://esphome.io/components/uart.html#uart
  - rx_pin: D4
    tx_pin: D3
    baud_rate: 9600
    id: senseair_s8_uart

  - rx_pin: D5
    baud_rate: 9600
    id: pms5003_uart

i2c:
  sda: D2
  scl: D1

sensor:
  - platform: pmsx003
    # https://esphome.io/components/sensor/pmsx003.html?highlight=pms5003
    type: PMSX003
    pm_1_0:
      name: "Particulate Matter <1.0µm Concentration"
      id: pm1
      filters:
        - sliding_window_moving_average:
            window_size: 30
            send_every: 30
    pm_2_5:
      name: "Particulate Matter <2.5µm Concentration"
      id: pm2
      filters:
        - sliding_window_moving_average:
            window_size: 30
            send_every: 30
    pm_10_0:
      name: "Particulate Matter <10.0µm Concentration"
      id: pm10
      filters:
        - sliding_window_moving_average:
            window_size: 30
            send_every: 30
    uart_id: pms5003_uart

  - platform: senseair
    # https://esphome.io/components/sensor/senseair.html
    co2:
      name: "SenseAir S8 CO2 Value"
      id: co2
    update_interval: 30s
    uart_id: senseair_s8_uart

  # - platform: sht3xd
  #   # https://esphome.io/components/sensor/sht3xd.html?highlight=sht31
  #   # This may not work as mine is not a -d model
  #   temperature:
  #     name: "AirGradient Temperature"
  #   humidity:
  #     name: "AirGradient Humidity"
  #   address: 0x44
  #   update_interval: 60s



font:
  - file: './fonts/arial.ttf'
    id: font1
    size: 24

display:
  - platform: ssd1306_i2c
    # https://esphome.io/components/display/ssd1306.html?highlight=display
    model: "SSD1306 64x48"
    id: oled_display
    reset_pin: D0
    address: 0x3C
    pages:
      - id: display_pm2
        lambda: |-
          it.print(0, 0, id(font1), "PM2");
          it.printf(64, 24, id(font1), TextAlign::TOP_RIGHT, "%.0f",id(pm2).state);
      - id: display_co2
        lambda: |-
          it.print(0, 0, id(font1), "CO2");
          it.printf(64, 24, id(font1), TextAlign::TOP_RIGHT, "%.0f",id(co2).state);

# cycle through display pages on a timer
interval:
  - interval: 5s
    then:
      - display.page.show_next: oled_display
      - component.update: oled_display

```

**Comment by MallocArray** [comment]
Experiencing the same issue when using a D1 Mini with the Ikea pm1006 sensor on UART D2, so it isn't limited to just the hardware of the AirGradient.
https://esphome.io/components/sensor/pm1006.html

Unfortunately, this sensor is new enough in ESPHome that I can't use the 1.18.0 release with it, so I just have to connect via USB currently to update the device until this is resolved and I can do OTA.

**Comment by mmakaay** [comment]
I also see issues with some of my devices, when using software UART. Incoming data on the RX pin disturb the ongoing OTA update, possibly because of interrupt handling for software UART. I haven't had the time to dig deep into this, but I want to compare the current code to the old code to see if anything was missed in the big ESP-IDF support update for 2021.10.0.

**Comment by tripplet** [comment]
I have the same problem with my device.
Maybe as a first workaround esphome could disable the software UART interrupt when an OTA upload starts + re enable it again if OTA fails.
There seem to be hooks available for all these OTA events: 
https://esphome.io/components/ota.html?highlight=ota#on-begin

There is a detach_interrupt() function for the *InternalGPIOPin* which is used inside the software UART component.
https://github.com/esphome/esphome/blob/68c854706752f387d51caf5adbf2d9091cf04b58/esphome/core/gpio.h#L85

But I'm not sure how to connect this to the OTA component because the InternalGPIOPin instance is protected.
So I see no easy way to do this via the yaml file.

**Comment by sanjayjuded** [comment]
I used ESPHome webpage on RPi4 directly on http://localhost:6052.  I could upload the code without any problems.  I then changed the update intervals of the sensors to 60s.  This was lower and it was causing a problem in OTA update.  Now I can update from another computer as I usually did before.

**Comment by MallocArray** [comment]
@sanjayjuded What type of sensors are you getting data from?  With my 2 different PM2.5 sensors, they are always sending data to the D1 mini that I can't control, I just use a sliding_window_moving_average filter to only send data up to HA on a less frequent basis, but the constant incoming data is what breaks my OTA with the 2021.10.x line.  I haven't seen any update to indicate it may be resolved in any newer version so I haven't upgraded yet.

**Comment by Doridian** [comment]
On local serial console, I get an exception when trying to OTA update (even updating via the WebUI fails with an exception)

```
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (0):
epc1=0x40226f80 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000

>>>stack>>>

ctx: sys
sp: 3fffe920 end: 3fffffb0 offset: 0190
3fffeab0:  000067e0 00000030 00000020 00000000
3fffeac0:  3ffefc54 00002972 00000020 401016da
3fffead0:  ffffffff 0000000e 00000000 40100ef9
3fffeae0:  e0031035 3fff2eed 00000000 40101721
3fffeaf0:  3fff3d24 00000010 00000020 00000022
3fffeb00:  3fffc200 40100e38 3fffc258 4000050c
3fffeb10:  400043e9 00000030 00000016 ffffffff
3fffeb20:  400044ab 3fffc718 3fffec10 08000000
3fffeb30:  60000200 08000000 00000000 00000000
3fffeb40:  0000ffff 00042035 00002035 0007d000
3fffeb50:  000005b4 00000000 00000001 00000030
3fffeb60:  0000000a 0000077b 0000077b 4010141b
3fffeb70:  00004690 000008d2 000008d2 4010141b
3fffeb80:  3fffebe0 3fffebf8 3fffebb0 00000006
3fffeb90:  3fffec7c 00000001 00000020 401016a3
3fffeba0:  3fffebe0 3fffebf8 3fff3d24 00000020
3fffebb0:  3fffec40 3fffebe0 3fffec7c 402221a1
3fffebc0:  3fffebec 3fffebe0 00000001 402221c0
3fffebd0:  3fffec7c 3fffebec 3fffebec 4021b379
3fffebe0:  3fff4300 0020002f 80000000 3fff4800
3fffebf0:  00000000 4000444e 3fff6900 009d009f
3fffec00:  00000002 4010660f 00000001 60000200
3fffec10:  00000002 4000410f 00001001 00000205
3fffec20:  3fffc718 40004a3c 0000007d 000005b4
3fffec30:  3fffc718 401068a0 0000007d 80000001
3fffec40:  40220b02 3ffef908 80000020 64707500
3fffec50:  00657461 80000005 3fff3c00 40221b9b
3fffec60:  80fe9c02 6e797300 70736563 80000000
3fffec70:  68707300 00656d6f 80fee828 000005b4
3fffec80:  00000000 000005b4 3ffef908 40221cb9
3fffec90:  3fff536c 00000120 3fff1158 00000000
3fffeca0:  00000000 00000001 3fff33f8 3fff469c
3fffecb0:  000005b4 3ffef908 3fff206c 4020e3a5
3fffecc0:  3fff469c 3fff1170 00000000 40218318
3fffecd0:  00000000 00000000 00000000 00000000
3fffece0:  3fff1158 00000000 3fff4c54 000005b3
3fffecf0:  00000000 3fff20ac 3fff469c 40227470
3fffed00:  00000000 3fff469c 3fff469c 4020e018
3fffed10:  3fff0078 3fff20ac 3fff469c 4020e090
3fffed20:  00000000 00000003 3fff0044 40239518
3fffed30:  3fff4780 0000eee0 3fff536c 000005b4
3fffed40:  0000f53b 3fff469c 3fff469c 402278f6
3fffed50:  00000000 3fff0050 3fff0044 00000557
3fffed60:  0000f4df 3fff469c 00000001 402172bd
3fffed70:  0000001a 000000d4 00000020 3ffee19c
3fffed80:  0000001a 3ffee828 00000020 401016a3
3fffed90:  40101936 000000ff 00000002 40101928
3fffeda0:  00000005 00000000 00000020 40100994
3fffedb0:  00000031 3fff2484 00000005 40102224
3fffedc0:  3ffe9c05 401054d7 3ffed640 000005b3
3fffedd0:  3fff4db2 000005b4 3fff469c 40218b95
3fffede0:  00007fff 02c562db 3ffee09c 40102f9c
3fffedf0:  00000000 4bc6a7f0 0000b59b 3fff4d6c
3fffee00:  00000001 000005b3 4bc6a7f0 00000000
3fffee10:  3ffea48c 7fffffff 40100d0e 3fff4d6c
3fffee20:  00000000 3fffee70 3fff45bc 40218d65
3fffee30:  00000000 3fffee70 3fff45bc 40225d86
3fffee40:  3fff449c 2c9f0300 4000050c 3fffc278
3fffee50:  40100000 3fffc200 00000022 3fff449c
3fffee60:  3fff4d6c 00000000 3fff45bc 40225f32
3fffee70:  3fff3bfc 3fff40c4 00000020 3ffefef8
3fffee80:  3fff0044 3fff00e4 3fff4d6c 3ffefef8
3fffee90:  3fff449c 3ffefedc 3ffefedd 40235cd4
3fffeea0:  00000000 3fffdcb0 3ffeef28 000005ec
3fffeeb0:  3fff0044 6603a8c0 00000020 401016da
3fffeec0:  00000000 000007f2 000007f2 3fff0044
3fffeed0:  3fff4d8a 3fff00e4 3fff0044 4023a029
3fffeee0:  3fff4d6c 00000014 00000000 401016a3
3fffeef0:  4024bd20 3fff40dc 3fff0784 3fff022c
3fffef00:  00000008 3fff0044 3fff4d6c 402331a1
3fffef10:  3fffdc80 3fff0784 3fff40dc 40232fc7
3fffef20:  402561de 3fff0784 3fff40dc 402561ef
3fffef30:  3fff4d7c 3fff4d6c 00000000 3fffdcb0
3fffef40:  4024b947 00000000 3fff40dc 402519f3
3fffef50:  40000f49 3fffdab0 3fffdab0 40000f49
3fffef60:  40000e19 00000005 0006d800 00000000
3fffef70:  00000000 aa55aa55 000000f7 401055dd
3fffef80:  401055e3 0006d800 00000000 3c26aba2
3fffef90:  4010000d 09d4113e 196aff79 d213556e
3fffefa0:  4024a828 3fffef3c 4024a7d9 3fffff18
3fffefb0:  3fffffc0 00000000 00000000 feefeffe
3fffefc0:  feefeffe feefeffe feefeffe feefeffe
3fffefd0:  feefeffe feefeffe feefeffe feefeffe
3fffefe0:  feefeffe feefeffe feefeffe feefeffe
3fffeff0:  feefeffe feefeffe feefeffe feefeffe
3ffff000:  feefeffe feefeffe feefeffe feefeffe
3ffff010:  feefeffe feefeffe feefeffe feefeffe
3ffff020:  feefeffe feefeffe feefeffe feefeffe
3ffff030:  feefeffe feefeffe feefeffe feefeffe
3ffff040:  feefeffe feefeffe feefeffe feefeffe
3ffff050:  feefeffe feefeffe feefeffe feefeffe
3ffff060:  feefeffe feefeffe feefeffe feefeffe
3ffff070:  feefeffe feefeffe feefeffe feefeffe
3ffff080:  feefeffe feefeffe feefeffe feefeffe
3ffff090:  feefeffe feefeffe feefeffe feefeffe
3ffff0a0:  feefeffe feefeffe feefeffe feefeffe
3ffff0b0:  feefeffe feefeffe feefeffe feefeffe
3ffff0c0:  feefeffe feefeffe feefeffe feefeffe
3ffff0d0:  feefeffe feefeffe feefeffe feefeffe
3ffff0e0:  feefeffe feefeffe feefeffe feefeffe
3ffff0f0:  feefeffe feefeffe feefeffe feefeffe
3ffff100:  feefeffe feefeffe feefeffe feefeffe
3ffff110:  feefeffe feefeffe feefeffe feefeffe
3ffff120:  feefeffe feefeffe feefeffe feefeffe
3ffff130:  feefeffe feefeffe feefeffe feefeffe
3ffff140:  feefeffe feefeffe feefeffe feefeffe
3ffff150:  feefeffe feefeffe feefeffe feefeffe
3ffff160:  feefeffe feefeffe feefeffe feefeffe
3ffff170:  feefeffe feefeffe feefeffe feefeffe
3ffff180:  feefeffe feefeffe feefeffe feefeffe
3ffff190:  feefeffe feefeffe feefeffe feefeffe
3ffff1a0:  feefeffe feefeffe feefeffe feefeffe
3ffff1b0:  feefeffe feefeffe feefeffe feefeffe
3ffff1c0:  feefeffe feefeffe feefeffe feefeffe
3ffff1d0:  feefeffe feefeffe feefeffe feefeffe
3ffff1e0:  feefeffe feefeffe feefeffe feefeffe
3ffff1f0:  feefeffe feefeffe feefeffe feefeffe
3ffff200:  feefeffe feefeffe feefeffe feefeffe
3ffff210:  feefeffe feefeffe feefeffe feefeffe
3ffff220:  feefeffe feefeffe feefeffe feefeffe
3ffff230:  feefeffe feefeffe feefeffe feefeffe
3ffff240:  feefeffe feefeffe feefeffe feefeffe
3ffff250:  feefeffe feefeffe feefeffe feefeffe
3ffff260:  feefeffe feefeffe feefeffe feefeffe
3ffff270:  feefeffe feefeffe feefeffe feefeffe
3ffff280:  feefeffe feefeffe feefeffe feefeffe
3ffff290:  feefeffe feefeffe feefeffe feefeffe
3ffff2a0:  feefeffe feefeffe feefeffe feefeffe
3ffff2b0:  feefeffe feefeffe feefeffe feefeffe
3ffff2c0:  feefeffe feefeffe feefeffe feefeffe
3ffff2d0:  feefeffe feefeffe feefeffe feefeffe
3ffff2e0:  feefeffe feefeffe feefeffe feefeffe
3ffff2f0:  feefeffe feefeffe feefeffe feefeffe
3ffff300:  feefeffe feefeffe feefeffe feefeffe
3ffff310:  feefeffe feefeffe feefeffe feefeffe
3ffff320:  feefeffe feefeffe feefeffe feefeffe
3ffff330:  feefeffe feefeffe feefeffe feefeffe
3ffff340:  feefeffe feefeffe feefeffe feefeffe
3ffff350:  feefeffe feefeffe feefeffe feefeffe
3ffff360:  feefeffe feefeffe feefeffe feefeffe
3ffff370:  feefeffe feefeffe feefeffe feefeffe
3ffff380:  feefeffe feefeffe feefeffe feefeffe
3ffff390:  feefeffe feefeffe feefeffe feefeffe
3ffff3a0:  feefeffe feefeffe feefeffe feefeffe
3ffff3b0:  feefeffe feefeffe feefeffe feefeffe
3ffff3c0:  feefeffe feefeffe feefeffe feefeffe
3ffff3d0:  feefeffe feefeffe feefeffe feefeffe
3ffff3e0:  feefeffe feefeffe feefeffe feefeffe
3ffff3f0:  feefeffe feefeffe feefeffe feefeffe
3ffff400:  feefeffe feefeffe feefeffe feefeffe
3ffff410:  feefeffe feefeffe feefeffe feefeffe
3ffff420:  feefeffe feefeffe feefeffe feefeffe
3ffff430:  feefeffe feefeffe feefeffe feefeffe
3ffff440:  feefeffe feefeffe feefeffe feefeffe
3ffff450:  feefeffe feefeffe feefeffe feefeffe
3ffff460:  feefeffe feefeffe feefeffe feefeffe
3ffff470:  feefeffe feefeffe feefeffe feefeffe
3ffff480:  feefeffe feefeffe feefeffe feefeffe
3ffff490:  feefeffe feefeffe feefeffe feefeffe
3ffff4a0:  feefeffe feefeffe feefeffe feefeffe
3ffff4b0:  feefeffe feefeffe feefeffe feefeffe
3ffff4c0:  feefeffe feefeffe feefeffe feefeffe
3ffff4d0:  feefeffe feefeffe feefeffe feefeffe
3ffff4e0:  feefeffe feefeffe feefeffe feefeffe
3ffff4f0:  feefeffe feefeffe feefeffe feefeffe
3ffff500:  feefeffe feefeffe feefeffe feefeffe
3ffff510:  feefeffe feefeffe feefeffe feefeffe
3ffff520:  feefeffe feefeffe feefeffe feefeffe
3ffff530:  feefeffe feefeffe feefeffe feefeffe
3ffff540:  feefeffe feefeffe feefeffe feefeffe
3ffff550:  feefeffe feefeffe feefeffe feefeffe
3ffff560:  feefeffe feefeffe feefeffe feefeffe
3ffff570:  feefeffe feefeffe feefeffe feefeffe
3ffff580:  feefeffe feefeffe feefeffe feefeffe
3ffff590:  feefeffe feefeffe feefeffe feefeffe
3ffff5a0:  feefeffe feefeffe feefeffe feefeffe
3ffff5b0:  feefeffe feefeffe feefeffe feefeffe
3ffff5c0:  feefeffe feefeffe feefeffe feefeffe
3ffff5d0:  feefeffe feefeffe feefeffe feefeffe
3ffff5e0:  feefeffe feefeffe feefeffe feefeffe
3ffff5f0:  feefeffe feefeffe feefeffe feefeffe
3ffff600:  feefeffe feefeffe feefeffe feefeffe
3ffff610:  feefeffe feefeffe feefeffe feefeffe
3ffff620:  feefeffe feefeffe feefeffe feefeffe
3ffff630:  feefeffe feefeffe feefeffe feefeffe
3ffff640:  feefeffe feefeffe feefeffe feefeffe
3ffff650:  feefeffe feefeffe feefeffe feefeffe
3ffff660:  feefeffe feefeffe feefeffe feefeffe
3ffff670:  feefeffe feefeffe feefeffe feefeffe
3ffff680:  feefeffe feefeffe feefeffe feefeffe
3ffff690:  feefeffe feefeffe feefeffe feefeffe
3ffff6a0:  feefeffe feefeffe feefeffe feefeffe
3ffff6b0:  feefeffe feefeffe feefeffe feefeffe
3ffff6c0:  feefeffe feefeffe feefeffe feefeffe
3ffff6d0:  feefeffe feefeffe feefeffe feefeffe
3ffff6e0:  feefeffe feefeffe feefeffe feefeffe
3ffff6f0:  feefeffe feefeffe feefeffe feefeffe
3ffff700:  feefeffe feefeffe feefeffe feefeffe
3ffff710:  feefeffe feefeffe feefeffe feefeffe
3ffff720:  feefeffe feefeffe feefeffe feefeffe
3ffff730:  feefeffe feefeffe feefeffe feefeffe
3ffff740:  feefeffe feefeffe feefeffe feefeffe
3ffff750:  feefeffe feefeffe feefeffe feefeffe
3ffff760:  feefeffe feefeffe feefeffe feefeffe
3ffff770:  feefeffe feefeffe feefeffe feefeffe
3ffff780:  feefeffe feefeffe feefeffe feefeffe
3ffff790:  feefeffe feefeffe feefeffe feefeffe
3ffff7a0:  feefeffe feefeffe feefeffe feefeffe
3ffff7b0:  feefeffe feefeffe feefeffe feefeffe
3ffff7c0:  feefeffe feefeffe feefeffe feefeffe
3ffff7d0:  feefeffe feefeffe feefeffe feefeffe
3ffff7e0:  feefeffe feefeffe feefeffe feefeffe
3ffff7f0:  feefeffe feefeffe feefeffe feefeffe
3ffff800:  feefeffe feefeffe feefeffe feefeffe
3ffff810:  feefeffe feefeffe feefeffe feefeffe
3ffff820:  feefeffe feefeffe feefeffe feefeffe
3ffff830:  feefeffe feefeffe feefeffe feefeffe
3ffff840:  feefeffe feefeffe feefeffe feefeffe
3ffff850:  feefeffe feefeffe feefeffe feefeffe
3ffff860:  00000005 00000000 00000020 40100994
3ffff870:  feefeffe feefeffe 00000005 40102224
3ffff880:  3ffe9c05 401054d7 3ffed690 feefeffe
3ffff890:  40102dbb 3ffed690 feefeffe feefeffe
3ffff8a0:  00000037 00631a0d 3ffee09c 40102f9c
3ffff8b0:  3ffea4b0 00000000 00000000 feefeffe
3ffff8c0:  00000006 00000007 3fff1ccc 40100335
3ffff8d0:  00000006 00000008 3fff1ccc 401003b2
3ffff8e0:  dd2afefa 00014e67 feefeffe feefeffe
3ffff8f0:  40100f11 0000000e 00000000 40100ef9
3ffff900:  e0031035 00000007 3fff1ccc 40100335
3ffff910:  00000000 00000008 3fff1ccc 00000022
3ffff920:  3fffc200 40100e38 3fffc258 4000050c
3ffff930:  4020c374 00000030 00000010 ffffffff
3ffff940:  40208dcc 3ffffe80 3fff228c 3ffffa80
3ffff950:  3ffffee8 4020c348 00000000 fffffffe
3ffff960:  000004a0 3fffc6fc 00000000 3fff221c
3ffff970:  3ffffe80 3ffffee8 3ffffa80 00000030
3ffff980:  3fffc200 40100e38 3fffc258 4000050c
3ffff990:  4022ee18 00000030 00000018 ffffffff
3ffff9a0:  4022f36f 00000001 00000001 00000014
3ffff9b0:  3ffffa90 3ffffad0 3ffffad4 3ffe8304
3ffff9c0:  00000090 00000005 3ffffc63 00000000
3ffff9d0:  3ff00000 3fff0604 000003ff 00000030
3ffff9e0:  4023f138 00000030 00000000 ffffffff
3ffff9f0:  00000046 00000007 3fff1e04 40100335
3ffffa00:  00000046 00000008 3fff1e04 401003b2
3ffffa10:  00000046 40180000 00000000 4022f316
3ffffa20:  00000003 00000002 00000000 40100ef9
3ffffa30:  3ffffa60 3ffffa64 3ffee09c 40102f9c
3ffffa40:  40100f11 03030000 00000000 40180000
3ffffa50:  00000000 00000001 3fff0625 3fff0fb4
3ffffa60:  00000000 3ffffe80 3fff0fb4 402093d0
3ffffa70:  00000000 3ffe8304 00000001 3fff0604
3ffffa80:  00000000 3cf30000 00000000 00000001
3ffffa90:  3fff0624 00000003 00000000 00000000
3ffffaa0:  00000000 00000000 00000000 00000000
3ffffab0:  00000000 00000000 00000000 40180000
3ffffac0:  3fff0625 00000006 3fff0625 00000000
3ffffad0:  3ffffc50 00000000 3ffffc50 4022f5ba
3ffffae0:  00000046 3ffffbf3 3ffe8304 4022a92e
3ffffaf0:  3ffffb68 3ffffb04 3ffffb00 00000001 <
3ffffb00:  4022f4f0 00000001 3ffffc50 4022f5ba
3ffffb10:  00000001 3fff0624 3ffffb60 00000000
3ffffb20:  00000000 00000000 0000000a 3fff0624
3ffffb30:  4022f4f0 3ffffbb0 3ffe8304 4022adac
3ffffb40:  3ffffb6c 3ffffb68 00000066 3ffffb64
3ffffb50:  00000000 3ffffc63 00000000 00000000
3ffffb60:  00000001 00000001 00000001 4022f500
3ffffb70:  3ffffc50 00000000 3ffe9730 00000001
3ffffb80:  00000001 00000066 00000001 00000001
3ffffb90:  3ffffcc0 00000001 3ffffcc0 3ffe8db6
3ffffba0:  00000000 3ffe8db5 3ffffc50 4022f9d5
3ffffbb0:  00000400 00000000 00000000 00000000
3ffffbc0:  00000001 00000001 00302066 4010141b
3ffffbd0:  00000001 00000006 00000001 00000017
3ffffbe0:  00000020 3fff47dc 3fff3c54 401018c8
3ffffbf0:  00000020 00000024 00000000 40180000
3ffffc00:  402028e0 00000004 00000001 3fff0a07
3ffffc10:  3ffffce0 3ffffcd0 00000018 3ffe8304
3ffffc20:  00003b20 00000009 00000001 40222226
3ffffc30:  00002db0 0000006d 00302064 3ffe8b41
3ffffc40:  00000008 3ffffd44 00000020 00000007
3ffffc50:  3ffefc54 00000010 00000020 401016da
3ffffc60:  00000002 000008e3 3fff40f4 00000000
3ffffc70:  00000001 3fff2ee4 000005ce 0000059a
3ffffc80:  0000000a 0000000a 3fff47dc 3fff45b6
3ffffc90:  00000786 3fff3af0 00000020 3fff3c0c
3ffffca0:  00004468 0000088d 0000088d 4010141b
3ffffcb0:  00000002 0000000d 0000000d 00000002
3ffffcc0:  3fff3c9c 0000000a 00000020 401016a3
3ffffcd0:  0000000b 00000036 3ffffd30 00000000
3ffffce0:  3ffffd20 00000002 00000000 40231f99
3ffffcf0:  0000000d 0000000d 00000019 4020200a
3ffffd00:  00000019 00000072 3ffffdc0 402022fc
3ffffd10:  00000000 4bc6a7f0 0000b1f0 00000019
3ffffd20:  00000000 00000000 4bc6a7f0 00000000
3ffffd30:  3fff45b4 3fff45b7 40100d0e 4978d4fd
3ffffd40:  00000000 3fff40f4 3ffffd90 40203ee4
3ffffd50:  00000000 00000000 3fff3b94 40100244
3ffffd60:  3fff3ba0 00000000 3fff3b94 4020135a
3ffffd70:  b3791a5d 40c00000 00000020 3fff3ce8
3ffffd80:  3ffffdc0 3fff3ba0 3ffffdc0 40204050
3ffffd90:  3fff3ba0 3fff4bda 3fff1094 3fff3ce8
3ffffda0:  3ffffdc0 3fff3ba0 3fff3b94 40226cde
3ffffdb0:  40c00000 3fff3b94 3fff1994 40201bac
3ffffdc0:  40228538 5d1a79b3 40c00000 00000000
3ffffdd0:  40c00000 3fff1994 3fff3ce4 40204bc1
3ffffde0:  3ffffe34 00000000 3fff1094 00000000
3ffffdf0:  3fff231c 40c00000 3fff230c 3fff3ce8
3ffffe00:  3fff3ca8 3ffffe5c 3fff3c9c 40202588
3ffffe10:  3fff1a78 00000000 3ffffe40 401000fd
3ffffe20:  00000000 4bc6a7f0 0000b58d 3fff3ce8
3ffffe30:  00000000 00000000 4bc6a7f0 00000000
3ffffe40:  3fff3c9c 00000000 40100d0e 83126e97
3ffffe50:  00000000 3ffffe70 00000010 00000000
3ffffe60:  3fff3ce8 000003e9 3fff3b94 40100244
3ffffe70:  3fff3ce8 000003e9 3fff3b94 402019b3
3ffffe80:  00000000 4bc6a7f0 0000b58d 00000003
3ffffe90:  00000000 00000000 4bc6a7f0 00000000
3ffffea0:  00000000 4bc6a7f0 40100d0e 947ae147
3ffffeb0:  00000000 00000000 4bc6a7f0 00000000
3ffffec0:  00000001 3ffef820 3fff2f50 4010079b
3ffffed0:  00000000 4bc6a7f0 0000b58d 3fff1c5c
3ffffee0:  00000000 00000000 4bc6a7f0 00000000
3ffffef0:  00000000 4bc6a7f0 40100d0e 9be76c8b
3fffff00:  40105ec1 00dd8e8a 3ffefb18 00000000
3fffff10:  3ffeef10 3ffefb18 00000010 3ffefb18
3fffff20:  0000b58d 3fff1c5c 40223398 3fffefa0
3fffff30:  3ffefb18 0000b58d 3fff1c5c 40224136
3fffff40:  3fff2e01 00000010 3ffef568 4010025c
3fffff50:  3fffdad0 00000010 3ffef568 40211abb
3fffff60:  3fff2e01 000004e0 3fff1500 3fff151c
3fffff70:  3fff2ed8 3fff12f4 3fff1c50 feefeffe
3fffff80:  00000000 00000000 00000001 3ffefa0c
3fffff90:  3fffdad0 00000000 3ffef9cc 40213e70
3fffffa0:  3fffdad0 00000000 3ffef9cc 402234bc
<<<stack<<<

--------------- CUT HERE FOR EXCEPTION DECODER ---------------
```

**Comment by mmakaay** [comment]
I'm afraid that exception backtrace is not usable without decoding it against the firmware.elf from your build output directory.

**Comment by Doridian** [comment]
@mmakaay Ah, fair enough. I ran through this with now a local copy of esphome which seems to have capabilities to decode the stack trace. I have attached that here.

The odd thing is that others in this thread also reported this with AirGradient DIY PCBs (which is what I am using).
And it started happening in the same version as this issue (if it is a separate issue)

I can flash fine with USB/Serial. I cannot flash via web or via OTA, both result in a stack trace like this one.

[exception.txt](https://github.com/esphome/issues/files/7629849/exception.txt)

**Comment by MallocArray** [comment]
I believe it has to do with a software version of UART due to the pins being used, which AirGradient chose to use on their PCB so it isn't easy to change.

I'm seeing similar behaviour with a D1 mini and an Ikea air quality monitor but I'm using the same pins and have the same type of device constantly sending data to the D1, so not all surprising that they are behaving the same way, but it at least rules out a single component causing the problem.

**Comment by Doridian** [comment]
Managed to figure out a fix (not sure if this is the correct fix): https://github.com/esphome/esphome/pull/2843/files

**Comment by shauneccles** [comment]
Using the dev version post the above PR appears to fix this issue on my two airgradient boards.

**Comment by DAVe3283** [comment]
Verified fixed in 2021.12.0!
